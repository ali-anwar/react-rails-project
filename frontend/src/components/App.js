import React from 'react';
import { Router, Route, Switch, Redirect } from 'react-router-dom';
import PageNotFound from './common/PageNotFound';
import Home from './landing/Home';
import createBrowserHistory from 'history/createBrowserHistory';
import HeaderNavContainer from './landing/HeaderNavContainer'; // eslint-disable-line import/no-named-as-default
import MealListContainer from './meal/MealListContainer';
import MealFormContainer from './meal/MealFormContainer';
import SignInContainer from './signIn/SignInContainer';
import SignUpContainer from './signUp/SignUpContainer';
import UserListContainer from './user/UserListContainer';
import UserFormContainer from './user/UserFormContainer';
import SettingFormContainer from './setting/SettingFormContainer';
import * as authActions from '../action/AuthActions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { REGULAR_ROLE, MANAGER_ROLE, ADMIN_ROLE } from '../lib/auth';

const history = createBrowserHistory();

export class App extends React.Component {
  render() {
    const { currentUser } = this.props;
    return (
      <div>
        <Router history={history}>
          <div>
            <HeaderNavContainer history={history} />

            <Switch>
              {currentUser && (
                <PrivateRoute
                  exact
                  path="/"
                  component={MealListContainer}
                  currentUser={currentUser}
                />
              )}
              {currentUser &&
                (currentUser.role === REGULAR_ROLE ||
                  currentUser.role === MANAGER_ROLE ||
                  currentUser.role === ADMIN_ROLE) && (
                  <PrivateRoute
                    exact
                    path="/meals"
                    component={MealListContainer}
                    currentUser={currentUser}
                  />
                )}
              {currentUser &&
                (currentUser.role === REGULAR_ROLE ||
                  currentUser.role === MANAGER_ROLE ||
                  currentUser.role === ADMIN_ROLE) && (
                  <PrivateRoute
                    path="/meals/new"
                    component={MealFormContainer}
                    currentUser={currentUser}
                  />
                )}
              {currentUser &&
                (currentUser.role === REGULAR_ROLE ||
                  currentUser.role === MANAGER_ROLE ||
                  currentUser.role === ADMIN_ROLE) && (
                  <PrivateRoute
                    path="/meal/:id"
                    component={MealFormContainer}
                    currentUser={currentUser}
                  />
                )}
              {currentUser &&
                (currentUser.role === MANAGER_ROLE ||
                  currentUser.role === ADMIN_ROLE) && (
                  <PrivateRoute
                    exact
                    path="/users"
                    component={UserListContainer}
                    currentUser={currentUser}
                  />
                )}
              {currentUser &&
                (currentUser.role === MANAGER_ROLE ||
                  currentUser.role === ADMIN_ROLE) && (
                  <PrivateRoute
                    path="/users/new"
                    component={UserFormContainer}
                    currentUser={currentUser}
                  />
                )}
              {currentUser &&
                (currentUser.role === REGULAR_ROLE ||
                  currentUser.role === MANAGER_ROLE) && (
                  <PrivateRoute
                    path="/settings"
                    component={SettingFormContainer}
                    currentUser={currentUser}
                  />
                )}
              {currentUser &&
                (currentUser.role === MANAGER_ROLE ||
                  currentUser.role === ADMIN_ROLE) && (
                  <PrivateRoute
                    path="/user/:id"
                    component={UserFormContainer}
                    currentUser={currentUser}
                  />
                )}
              <Route exact path="/" component={Home} />

              <Route
                path="/login"
                render={props =>
                  !currentUser ? (
                    <SignInContainer {...props} />
                  ) : (
                    <Redirect
                      to={{
                        pathname: '/',
                        state: { from: props.location }
                      }}
                    />
                  )
                }
              />
              <Route
                path="/signup"
                render={props =>
                  !currentUser ? (
                    <SignUpContainer {...props} />
                  ) : (
                    <Redirect
                      to={{
                        pathname: '/',
                        state: { from: props.location }
                      }}
                    />
                  )
                }
              />
              {!currentUser ? <Redirect to="/" /> : null}
              <Route component={PageNotFound} />
            </Switch>
          </div>
        </Router>
      </div>
    );
  }
}

const PrivateRoute = ({ currentUser, component: Component, ...rest }) => {
  return (
    <Route
      {...rest}
      render={props => {
        return currentUser ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{
              pathname: '/login',
              state: { from: props.location }
            }}
          />
        );
      }}
    />
  );
};

const mapStateToProps = state => ({
  currentUser: state.userReducer.currentUser
});

const mapDispatchToProps = dispatch => ({
  action: bindActionCreators({ ...authActions }, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);
