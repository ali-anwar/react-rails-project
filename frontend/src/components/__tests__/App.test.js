import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {App} from '../App';


describe('App.test.js', () => {
  const props = {
    location: 'https://caloriessystem.com'
  };

  it('renders as expected', () => {
    const wrapper = shallow(<App {...props} />);
    const tree = toJson(wrapper);
    
    expect(wrapper.length).toEqual(1);
    expect(tree).toMatchSnapshot();
  });
});
