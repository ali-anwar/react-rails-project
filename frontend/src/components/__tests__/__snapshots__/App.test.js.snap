exports[`App.test.js renders as expected 1`] = `
<div>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }>
    <div>
      <Connect(HeaderNavContainer)
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        } />
      <Switch>
        <Route
          component={[Function]}
          exact={true}
          path="/" />
        <Route
          path="/login"
          render={[Function]} />
        <Route
          path="/signup"
          render={[Function]} />
        <Redirect
          push={false}
          to="/" />
        <Route
          component={[Function]} />
      </Switch>
    </div>
  </Router>
</div>
`;
